test/periodic.jl

```diff
@@ -0,0 +1,332 @@
+@testset "periodic" begin
+    # Constructors
+    @test Periodic(sin, 5) == Periodic(sin, 5, 0)
+
+    @testset "_periodic" begin
+        @testset "$f" for f in (sin, cos)
+            # A typical 24-hour day
+            for h in 0:24
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 6, 1, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 24 * h)
+                @test result ≈ expected atol=1e-15
+            end
+
+            # A spring daylight-saving time change where the day has 23-hours
+            for h in 0:23
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 3, 8, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 23 * h)
+                @test result ≈ expected atol=1e-15
+            end
+
+            # A fall daylight-saving time change where the day has 25-hours
+            for h in 0:25
+                result = _periodic(
+                    f,
+                    ZonedDateTime(2015, 11, 1, tz"America/Winnipeg") + Hour(h),
+                    Day(1),
+                )
+                expected = f(2π / 25 * h)
+                @test result ≈ expected atol=1e-15
+            end
+        end
+
+        @testset "negative period" begin
+            @test ==(
+                _periodic(sin, DateTime(2000, 1, 1, 3), Day(-1)),
+                -_periodic(sin, DateTime(2000, 1, 1, 3), Day(1)),
+            )
+
+            @test ==(
+                _periodic(cos, DateTime(2000, 1, 1, 3), Day(-1)),
+                _periodic(cos, DateTime(2000, 1, 1, 3), Day(1)),
+            )
+        end
+
+        @testset "phase shift" begin
+            @test _periodic(sin, DateTime(1), Week(1)) == 0
+            @test _periodic(sin, DateTime(1), Week(1), Day(7)) == 0
+            @test _periodic(sin, DateTime(1), Week(1), Day(-7)) == 0
+
+            # Using the phase shift we can make a Thursday the beginning of the week.
+            # Note: that these calculations can change depending on the period type given.
+            @test _periodic(sin, DateTime(1970), Week(1), Day(3)) == 0
+            @test _periodic(sin, DateTime(1970), Second(Week(1)), Day(5)) == 0
+
+            # Phase shifts larger than the period should not change the result
+            @test ==(
+                _periodic(sin, DateTime(1970), Day(1), Day(0)),
+                _periodic(sin, DateTime(1970), Day(1), Day(5)),
+            )
+        end
+    end
+
+    @testset "Real input" begin
+        @testset "$f" for f in (sin, cos)
+            p = Periodic(f, 5, 2)
+
+            @test p isa Transform
+
+            @testset "Vector" begin
+                x = collect(0.:11.)
+                expected = f.(2π .* (x .- 2) ./ 5)
```

Comment:
can you hard-code this? this is essentially the same code in the function being tested. So if there's a bug in that (unlikely, but still) we won't catch it here. Same for any other tests that call `f.(2π .* (x .- x') ./ d)`

Reply:
Doing this, for NamedTuple it looks like I need to use `collect(nt) ≈ collect(nt_expected) atol=1e-14` since there is no `isapprox` for NamedTuples.

I might be able to adjust the hard-coded values to match the output, so that we can use `==` (initially I got them from applying sin/cos in the REPL), but I'm not sure if that will be stable.

Reply:
>I might be able to adjust the hard-coded values to match the output,

I wouldn't bother tbh. rounding errors are to be expected on `sin` / `cos` so `isapprox` and checking the type is fine